### PR TITLE
ARROW-13160: [CI][C++] Use binary caching for vcpkg builds

### PR DESCRIPTION
**Pull request by assignUser**

### CODE REVIEW

**Comment by assignUser** [comment]
@github-actions submit wheel-macos-*-amd64 test-build-vcpkg-win

**Comment by assignUser** [comment]
@github-actions crossbow submit wheel-macos-*-amd64 test-build-vcpkg-win

**Comment by raulcd** [comment]
@github-actions crossbow submit --group verify-rc-source --group verify-rc-binaries --group verify-rc-wheels

**Comment by raulcd** [comment]
The above crossbow comment was not expected to be run on this PR. Sorry about triggering all these jobs.

**Comment by assignUser** [comment]
@github-actions crossbow submit wheel-macos-mavericks-cp39-amd64

**Comment by assignUser** [comment]
@github-actions crossbow submit wheel-macos-mavericks-cp39-amd64

**Comment by assignUser** [comment]
This reduces the run time by about 50 minutes (-60%) 🚀 

I have not added it to the M1 workflows as those already have a persistent cache through the self-hosted runner.
I have tested building the packages without the separate `openssl` installation step. The issue seems to have been resolved upstream.

@raulcd @kszucs

File: dev/tasks/python-wheels/github.osx.amd64.yml [hunk 1]
```diff
@@ -68,10 +63,23 @@ jobs:
...
-      - name: Install OpenSSL
-        run: vcpkg install openssl
-
+      - name: Setup NuGet Credentials
+        shell: bash
+        env: 
+          GITHUB_TOKEN: {{ '${{ secrets.GITHUB_TOKEN }}' }}
+        run: |          
+          mono `vcpkg fetch nuget | tail -n 1` \
+          sources add \
```
Comment (kou) [inline dev/tasks/python-wheels/github.osx.amd64.yml:72]: Could you indent continuous lines?

```bash
mono ... \
  sources add \
  ...
```

File: dev/tasks/python-wheels/github.osx.amd64.yml [hunk 2]
```diff
@@ -68,10 +63,23 @@ jobs:
...
+          -source "https://nuget.pkg.github.com/ursacomputing/index.json" \
```
Comment (kou) [inline dev/tasks/python-wheels/github.osx.amd64.yml:73]: Can we use `${{ github.repository_owner }}` instead of `ursacomputing`?

Comment (assignUser) [reply dev/tasks/python-wheels/github.osx.amd64.yml:73]: I didn't do this as to not pollute peoples repos with nuget packages but I guess this is an on demand job so it shouldn't be an issue :)

Comment (assignUser) [reply dev/tasks/python-wheels/github.osx.amd64.yml:73]: I used the envvar version, is cleaner with jinja

**Comment by kszucs** [comment]
@assignUser really great, thanks for working on this!

Could we incorporate similar changes to the manylinux and windows docker builds as well with a baked in crossbow repository? Of course this can be a follow-up work.

**Comment by assignUser** [comment]
@kszucs Thanks! I'll  take a look, we will need to pass in the token but that shouldn't be an issue 🚀

File: dev/tasks/python-wheels/github.osx.amd64.yml (hunk 2 above)
Comment (assignUser) [reply dev/tasks/python-wheels/github.osx.amd64.yml]: Ah of course sorry

**Review by kou — APPROVED** [review]
+1

Thanks!

**Pull request merged**